python/cuda_parallel/cuda/parallel/experimental/_utils/protocols.py

```diff
@@ -14,15 +14,29 @@
 from ..typing import DeviceArrayLike
 
 
-def get_dtype(arr: DeviceArrayLike) -> np.dtype:
-    typestr = arr.__cuda_array_interface__["typestr"]
+def get_data_pointer(arr: DeviceArrayLike) -> int:
+    try:
+        # TODO: this is a fast path for CuPy until
+        # we have a more general solution.
+        return arr.data.ptr  # type: ignore
+    except AttributeError:
+        return arr.__cuda_array_interface__["data"][0]
 
-    if typestr.startswith("|V"):
-        # it's a structured dtype, use the descr field:
-        return np.dtype(arr.__cuda_array_interface__["descr"])
-    else:
-        # a simple dtype, use the typestr field:
-        return np.dtype(typestr)
+
+def get_dtype(arr: DeviceArrayLike) -> np.dtype:
+    try:
+        # TODO: this is a fast path for CuPy until
+        # we have a more general solution.
+        return np.dtype(arr.dtype)  # type: ignore
+    except Exception:
+        typestr = arr.__cuda_array_interface__["typestr"]
+
+        if typestr.startswith("|V"):
+            # it's a structured dtype, use the descr field:
+            return np.dtype(arr.__cuda_array_interface__["descr"])
```

Comment:
In certain cases we might need to access `arr.__cuda_array_interface__` more than once. Given that it is expensive, why not save it to a variable and reuse the interface dictionary.

Reply:
See above discussion: https://github.com/NVIDIA/cccl/pull/3718#discussion_r1945070697